File: piece_classes/groped_pieces/Direction.cpp
//
// Created by luca on 4/20/23.
//

#include "Direction.h"

Direction operator-(Direction direction){
    switch (direction) {
        case UP:
            return DOWN;
        case DOWN:
            return UP;
        case RIGHT:
            return LEFT;
        case LEFT:
            return RIGHT;
        default:
            throw std::invalid_argument("unknown direction");
    }
}
std::ostream& operator<<(std::ostream & os, Direction direction){
    switch (direction) {
        case UP:
            return os << "UP";
        case DOWN:
            return os << "DOWN";
        case RIGHT:
            return os << "RIGHT";
        case LEFT:
            return os << "LEFT";
        default:
            throw std::invalid_argument("unknown direction");
    }
}

File: cpp_lib/rust_solving/src/piece_group/direction.rs
// the 4 direction a piece can be compared on

use crate::piece_group::Direction::{DOWN, LEFT, RIGHT, UP};

#[derive(Clone,Copy)]
pub enum Direction {
    UP = 0,
    RIGHT = 1,
    DOWN = 2,
    LEFT = 3
}

impl std::ops::Neg for Direction{
    type Output = Direction;
    fn neg(self) -> Self::Output {
        match self {
            UP => DOWN,
            DOWN => UP,
            LEFT => RIGHT,
            RIGHT => LEFT
        }
    }
}